[RFC] OPEA Inference Microservices (OIM) (#327)
opea-project/docs: community/rfcs/25-03-07-OPEA-001-OIM-Operator.md

# RFC: OPEA Inference Microservices (OIM)

This RFC talks about creating OPEA inference microservices (OIM) similar to NVIDIA NIMs.

## Author(s)

[Poussa, Sakari](https://github.com/poussa),
[Bhandaru, Malini](https://github.com/mkbhanda),
[Lehtonen, Markus](https://github.com/marquiz),
[Kanevsky, Alexander](https://github.com/kad),
[Lu, Lianhao](https://github.com/lianhao),
[Yang, Yuhan](https://github.com/PeterYang12)

## Status

`Under Review`

## Objective

OPEA seeks to ease enterprise GenAI adoption in a landscape that is fast moving, has skills gaps, and is cost conscious. A look at OPEA's GenAIExamples illustrates the choices a user must make: whether to use Docker or Kubernetes, the model serving framework, whether to optimize for latency or throughput or quantization, in addition to the hardware they want to run on. Can we abstract some of this complexity for inference services?

> Nvidia's NIM does just this. It expects a user to only specify a model and what they would like to optimize for, their key performance indicator (KPI), be it low latency, high throughput or quantization. Nvidia abstracts away both the need to choose a model serving framework and determining the optimum parameters for the same to achieve the KPIs on provided hardware.

Author: fixed.

## Motivation

> Nvidia delivers this ease of use through tooling that first detects the runtime hardware, which is then used to filter for profiles given the desired model and KPIs. A profile is a recipe specifying how to launch an inference service.  It specifies not only which model server framework (such as TGI or vLLM or Ollama ) to use but also number of GPU cards, amount of memory, degree of tensor parallelism, batch size etc. Lastly the profile is used to launch the inference service on the given hardware.

Author: fixed.

> The profiles are established through offline experimentation using tools like MLPerf. For a model, one or more model frameworks are explored on different hardware to determine settings that deliver on the desired KPIs. Constructing profiles is non-trivial, resource intensive work.

Author: fixed.

## Design Proposal

> We seek to offer a similar ease of use experience in OPEA for the inference services. Nvidia’s open source NIM Operator is a tool that eases launching inference services. OPEA plans to extend it to deliver an Operator that provides similar functionality but also supports multi-vendor hardware and shall endeavor to work with Kubernetes and its flavors such as Red Hat OpenShift and SUSE Rancher, and their AI variants by working with our partners and potentially integrating with platforms such as KServe and RayAI.  The first release will provide minimal functionality and over time with community contributions grow richer.  We anticipate providers of AI platforms that operate on a multitude of hardware platforms will find the HW optimized profiles valuable in providing cost-effective performant inference services.

Reviewer: Thank you @lianhao and @eero-t! Perhaps these liveness and readiness probes times could be a function of (HW + model-size) in a busy cluster.

> We seek to offer a similar ease of use experience in OPEA for the inference services. Nvidia’s open source NIM Operator is a tool that eases launching inference services. OPEA plans to extend it to deliver an Operator that provides similar functionality but also supports multi-vendor hardware and shall endeavor to work with Kubernetes and its flavors such as Red Hat OpenShift and SUSE Rancher, and their AI variants by working with our partners and potentially integrating with platforms such as KServe and RayAI.  The first release will provide minimal functionality and over time with community contributions grow richer.  We anticipate providers of AI platforms that operate on a multitude of hardware platforms will find the HW optimized profiles valuable in providing cost-effective performant inference services.

Reviewer 2: Kubernetes deployment is good and scalable, but it is a little complex to setup the kubernets environments. Shall we also support such enhancements with docker image, like NIM, ollama, to deploy simply, efficient, complete, optimized service?

Reviewer: Indeed we want the OIM container images to be runnable in a docker environment, much like what is discussed in the topmost section in https://developer.nvidia.com/blog/a-simple-guide-to-deploying-generative-ai-with-nvidia-nim/.

Further to @ftian1 architecture question, once again pointing to an Nvidia blog: https://developer.nvidia.com/blog/managing-ai-inference-pipelines-on-kubernetes-with-nvidia-nim-operator/, see Figure 1. Our OPEA GenAIExample pipelines would use the OIM just like an OpenAI style API call. We also have the Langchain and LlamaIndex type integrations.

> With regard to profiles we shall start with supporting vLLM as the model framework and build profiles for a few popular models. Profiles may also contain Kubernetes information (e.g., resource requests) to ease deployments. We do anticipate constant enhancements and new offerings in the model serving space and a steady influx of new models.

Reviewer 3: Accurate resource requests, probe & shutdown timings do not just "ease" deployment, they are needed to guarantee that services actually work as intended when they are stressed / in resource constrained situations (= cost-optimized production setups).  Especially with autoscaling, but also otherwise.

=> I would phrase this more strongly, for example:
```suggestion
With regard to profiles we shall start with supporting vLLM as the model framework and build profiles for a few popular models. Profiles may also contain more accurate resource usage information for Kubernetes deployments (allowing better k8s scheduling decisions and preventing failures on resource contention). We do anticipate constant enhancements and new offerings in the model serving space and a steady influx of new models.
```

The work will entail 

> 1) Auto-detecting hardware type and what is allocated.

Reviewer 4: Sakari, 

The overall description is ok for me, just wonder will we have a high level architecture diagram to show how it interactives with other OPEA components? And how auto-detecting and allocating h/w works?

Author: @ftian1 now an architecture diagram is included.

2) Taking the user specified model and profile (if not specified using a default profile) to retrieve configurations that are possible on the allocated hardware type and quantity. For instance, a configuration that leverages tensor parallelism is infeasible if only one GPU card is allocated.

3) Automating performance studies to identify optimum settings for a given hardware with various resource allocations for a given model and constructing profile specific configurations. These studies will be repeated to take advantage of new models, frameworks, and other enhancements.

4) Providing an open-source tool, such as [ORAS](https://oras.land/), to discover and list available profiles/configurations for a model.

> 5) Publishing the profiles and other artifacts to an OCI compatible registry, such as Docker Hub or ghcr.io

Author: fixed.

## Author(s)

[Poussa, Sakari](https://github.com/poussa),
[Bhandaru, Malini](https://github.com/mkbhanda),
[Lehtonen, Markus](https://github.com/marquiz),
[Kanevsky, Alexander](https://github.com/kad),
[Lu, Lianhao](https://github.com/lianhao),
[Yang, Yuhan](https://github.com/PeterYang12)

## Status

`Under Review`

## Objective

OPEA seeks to ease enterprise GenAI adoption in a landscape that is fast moving, has skills gaps, and is cost conscious. A look at OPEA's GenAIExamples illustrates the choices a user must make: whether to use Docker or Kubernetes, the model serving framework, whether to optimize for latency or throughput or quantization, in addition to the hardware they want to run on. Can we abstract some of this complexity for inference services? 

Nvidia's NIM does just this. It expects a user to only specify a model and what they would like to optimize for, their key performance indicator (KPI), be it low latency, high throughput or quantization. Nvidia abstracts away both the need to choose a model serving framework and determining the optimum parameters for the same to achieve the KPIs on provided hardware.  

## Motivation

> Nvidia delivers this ease of use through tooling that first detects the runtime hardware, which is then used to filter for profiles given the desired model and KPIs. A profile is a recipe specifying how to launch an inference service.  It specifies not only which model server framework (such as TGI or vLLM or Ollama ) to use but also number GPU cards, amount of memory, degree of tensor parallelism, batch size etc. Lastly the profile is used to launch the inference service on the given hardware. 

Reviewer: ```suggestion
Nvidia delivers this ease of use through tooling that first detects the runtime hardware, which is then used to filter for profiles given the desired model and KPIs. A profile is a recipe specifying how to launch an inference service.  It specifies not only which model server framework (such as TGI or vLLM or Ollama ) to use but also number of GPU cards, amount of memory, degree of tensor parallelism, batch size etc. Lastly the profile is used to launch the inference service on the given hardware. 
```

Author: fixed.

> The profiles are established through offline experimentation using tools like MLPerf. For a model, one or more model frameworks are explored on different hardware to determine settings the deliver on the desired KPIs. Constructing profiles is non-trivial, resource intensive work.  

Reviewer: ```suggestion
The profiles are established through offline experimentation using tools like MLPerf. For a model, one or more model frameworks are explored on different hardware to determine settings that deliver on the desired KPIs. Constructing profiles is non-trivial, resource intensive work.  
```

Author: fixed.

## Design Proposal

> We seek to offer a similar ease of use experience in OPEA for the inference services. Nvidia’s open source NIM Operator is a tool that eases launching inference services. OPEA plans to extend it to deliver an Operator that provides similar functionality but also supports multi-vendor hardware and shall endeavor to work with Kubernetes and flavors such as Red Hat OpenShift and SUSE Rancher, their AI variants by working with our partners and potentially integrating with platforms such as KServe and RayAI.  The first release will provide minimal functionality and over time with community contributions grow richer.  We anticipate providers of AI platforms that operate on a multitude of hardware platforms will find the HW optimized profiles valuable in providing cost-effective performant inference services.

Reviewer: ```suggestion
We seek to offer a similar ease of use experience in OPEA for the inference services. Nvidia’s open source NIM Operator is a tool that eases launching inference services. OPEA plans to extend it to deliver an Operator that provides similar functionality but also supports multi-vendor hardware and shall endeavor to work with Kubernetes and its flavors such as Red Hat OpenShift and SUSE Rancher, and their AI variants by working with our partners and potentially integrating with platforms such as KServe and RayAI.  The first release will provide minimal functionality and over time with community contributions grow richer.  We anticipate providers of AI platforms that operate on a multitude of hardware platforms will find the HW optimized profiles valuable in providing cost-effective performant inference services.
```

Author: fixed.

With regard to profiles we shall start with supporting vLLM as the model framework and build profiles for a few popular models. We do anticipate constant enhancements and new offerings in the model serving space and a steady influx of new models.

The work will entail 

> 1) auto-detecting hardware and what is allocated

Reviewer: ```suggestion
1) Auto-detecting hardware type and what is allocated.
```

Author: fixed.

2) Taking the user specified model and profile (if not specified using a default profile) to retrieve configurations that are possible on the allocated hardware type and quantity. For instance, a configuration that leverages tensor parallelism is infeasible if only one GPU card is allocated. 

> 3) Automating performance studies to identify optimum settings for a given hardware with various resource allocations for a given model and constructing profile specific configurations. These studies will be repeated to take advantage of new models, frameworks and other enhancements.  

Reviewer: ```suggestion
3) Automating performance studies to identify optimum settings for a given hardware with various resource allocations for a given model and constructing profile specific configurations. These studies will be repeated to take advantage of new models, frameworks, and other enhancements.  
```

Author: fixed.

4) Providing a tool to discover and list available profiles/configurations for a model. Using this tool in the OPEA Operator, and within our inference microservices to identify relevant configurations and use. 

> 

Reviewer: Could 4 & 5 be merged, as in use ORAs from within the Operator, container ..?

Author: done.

5) Leveraging open source tools like ORAS `oras-project/oras` to retrieve profiles and other artifacts

> 6) Publishing the profiles and other artifacts to OCI registry, such as Docker Hub or ghcr.io 

Reviewer: ```suggestion
6) Publishing the profiles and other artifacts to an OCI compatible registry, such as Docker Hub or ghcr.io. 
```

## Alternatives Considered

> Creating everything from scratch was rejected becase an open-source alternative exists and provides most of the features OPEA project needs.

Reviewer: ```suggestion
Creating everything from scratch was rejected because an open-source alternative exists and provides most of the features OPEA project needs.
```

Author: fixed.
